### ISSUE TITLE
XDG_RUNTIME_DIR is mislabelled, systemd-user PAM stack lacks pam_selinux, "systemctl edit" should label created files properly

### ISSUE THREAD

**Issue by ghost** [systemd/systemd#257]
Back in the old day's SELinux was transparent to most of the user space, including init. This was possible because init was relatively simple compared to systemd.

With the advent of systemd, we were forced to make systemd SELinux aware. Mainly because of systemds' unit concept compared to inits' rc script concept.

Initial systemd SELinux support was implemented reasonably well. However, it is proving to be hard to maintain. Mainly because in order for systemd to be SELinux aware, it's developers should ideally be SELinux aware as well. They would ideally be able to identify where systemd should manually intervene to ensure optimal compatibility with SELinux.

Unfortunately this turns out to not be the case, and so every once in a while i feel that i have to play the "systemd SELinux awareness boogie-man" in order to keep things in-sync so it seems.

Here are a couple of "notes" for your kind consideration:
1. The `systemctl edit` functionality should, probably, use the SELinux `setfscreatecon()` functionality to ensure that `overrides.conf` files are created with the SELinux security context that was specified by the administrators' SELinux security policy. This would also apply to any unit files potentially created in the future with the `systemctl edit` functionality.
2. The `/etc/pam.d/systemd-user` pam configuration file should, probably, enclose calls to `pam_selinux` in it's stack to ensure that systemd user instances are associated with the SELinux security context that was specified by the administrators' SELinux security policy.
3. The systemd logind daemon should, probably, create `/run/user/$UID` using the SELinux `setfscreatecon()` functionality to ensure that `$XDG_RUNTIME_DIR` is created with the security context that was specified by the administrators' SELinux security policy.

In addition, the systemd logind daemon should, probably, use the `lsetfilecon()` SELinux functionality, as soon as it mounted tmpfs on each `$XDG_RUNTIME_DIR`, to ensure that the security context that was used to label `/run/user/$UID` when it was created, get's re-applied to the mountpoint after tmpfs was mounted to it.

**Comment by poettering** [comment]
I'd prefer if seperate issues were filed for each of the issues you point out. But OK.

Regarding your second issue: we just include "system-auth" from that file. Why isn't pam_selinux included in that? Given that selinux is effectively a fedora/redhat-only implementation detail it appears to me we shouldn't mention the PAM module upstream, but rely that "system-auth" includes it.

**Comment by poettering** [comment]
Please check if #259 fixes the third issue for you.

**Comment by ghost** [comment]
I believe that you are wrong when you say that SELinux is effectively a Fedora/Redhat only implementation. Just because SELinux is optional with some distributions does not make it any less effective implementation in my view. (Gentoo, Debian, CoreOS to name a few)

The pam_selinux module is not included with the system-auth stack because it would be inappropriate i believe. Not all callers of the system-auth stack need to be able to set security contexts on processes that they fork.

**Comment by poettering** [comment]
I don't see how it would make sense to run a PAM session without setting up selinux for it. Please get feedback from Daniel Walsh or so about this.

**Comment by ghost** [comment]
@mgrepl @rhatdan would you be so kind to provide some input here?

**Comment by ghost** [comment]
As soon as any of these changes land in Fedora Rawhide i will be able to test them and provide feedback.

**Comment by rhatdan** [comment]
@poettering We don't want pam_selinux called for su and sudo, just for initial SELinux login sessions, which is why it is not in system-auth, as I recall.

Not sure what systemd-user or systemd-logind are creating login user processes.  If they are then they should be calling pam_selinux.  If they are called after the login program then they should just continue with the default label.

For the most part we want user sessions to be labeled as unconfined_t, or staff_t, user_t, guest_t ...

**Comment by rhatdan** [comment]
Mounting of the tmpfs on  /run probably could be handled using a context mount point rather then a setfscreatecon.

Something like mount -o rootcontext='system_u:object_r:user_tmp_t:s0"  Or whatever the default lable is.

**Comment by ghost** [comment]
@rhatdan Would the -o rootcontext='' value be customizable by the administrator?

Also have you ever used the systemd-user instance in a meaningful way? I am pretty sure that it's SELinux support is currently very broken.

**Comment by ghost** [comment]
I have been using pam_selinux for systemd-user for about six months now and i am confident that the systemd-user daemon context needs to be customizable by the administrator. The way to achieve this is through pam_selinux.

If you give me a chance to demonstrate my case then i am confident that i can make the case. At least i think i will be able to demonstrate that it is currently broken.

**Comment by poettering** [comment]
I am happy to take a patch that adds pam_selinux to our PAM snippet. I am not sure I can test this on my own, and I grok selinux too little to write this myself.

@rhatdan note that these are _per-user_ tmpfs instances in /run/user/<user>. Are you suggesting there should be code that reads the appropriate label from the database and then passes it as mount option instead of applying it afterwards? In that case I'd need a patch, my selinux-fu is too limited. I did prepare #259 though, which first mounts and then alters the label.

**Comment by ghost** [comment]
You are making a valid point @poettering. The systemd people do not understand SELinux fully and the SELinux people do not understand systemd fully.

The contexts of the `XDG_RUNTIME_DIR` and of the `systemd user` instance need to be customizable by the administrator. I have been using this for quite some time and I believe i have it right.

I'd be glad to some how demonstrate my point to both systemd as well as SELinux people.

**Comment by ghost** [comment]
A terrible attempt to explain by example some of the issues involved: 

https://www.youtube.com/watch?v=1HR-RGFtqTw

The above is not comprehensive! It focuses on identities but identities are just a means to a end. It is important never the less.

**Comment by ghost** [comment]
Say what you will but at least I am trying: http://doverride.blogspot.nl/2015/06/trying-to-make-case.html

**Comment by mgrepl** [comment]
I understand your points. 
- /run/user/<UID> labeling
  Yes it needs to set a context defined by the policy. Just not sure what the best solution is here. So we have "mount option" vs.  "applying it afterwards". I like to use a mount option with a context which matches the policy.
- pam_selinux + systemd --user
  Yes, with the current setup we have it running under wrong SELinux context. What I see as a problem is you would need to have

allow userdomain init_exec_t:file entrypoint;

if I am correct together with 

allow init_t userdomain:process transtiion;

which is not so big deal.

**Comment by ghost** [comment]
I believe @poettering has the optimal approach here https://github.com/systemd/systemd/pull/259 (provided that the code works, which i will test as soon as it lands in Fedora Rawhide)

It is simple and effective.

The only potential problem might be that there is probably a small time window for a race. Between the time tmpfs in mounted on /run/user/$UID and when lsetfilecon() actually reset the context

**Comment by mgrepl** [comment]
Yeap. I am also fine with #259.

And how does your policy look related to pam+systemd-user? The problem what I see we have a check on /usr/lib/systemd/systemd which is caused by pam_selinux.

**Comment by ghost** [comment]
@mgrepl my advice to you with regard to adapting the policy, at least initially, is as you state above
1. focus on unconfined users initially, when that works shift focus to confined users. (use rawhide for this testing)
2. simply start by adding pam_selinux to the /etc/pam.d/systemd-user stack. (i am using the nottys option with the pam_selinux module, see man:pam_selinux)
3. edit /etc/selinux/targeted/contexts/users/unconfined_u and add something like the following: system_r:init_t:s0 unconfined_r:unconfined_t:s0-s0:c0.c1023
4. Then test this in permissive mode. Ensure that the users systemd --user instance runs with the unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 context associated with it just like (most of) the other processes listed with ps -xZ
5. modify the security policy using any avc denials and other SELinux events in the audit.log/journal.

I would initially focus on a transition on unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023

This is because eventually you may want systemd --user instances for confined users associated with a private systemd user context, but that is something you will have to decide for yourself.

Since this has been broken for a long time, you can take your time to get it right. Just deal with default unconfined logins because that is the most urgent and then explore your possibilities for other confined logins

**Comment by ghost** [comment]
@mgrepl This is what i have for unconfined users:

(optional systemd_optional_unconfineduser
        (call unconfineduser_noatsecure_all_subjects (systemd_t))
        (call unconfineduser_send_sigchld_all_subjects (systemd_t))
        (call unconfineduser_create_keyring_all_subjects (systemd_t))
        (call unconfineduser_shell_exec_manual_subject_type_transition_all_subjects (systemd_t)))

Note the noatsecure permission.

https://github.com/doverride/laptop/blob/master/sources/modules/contrib/system/systemd/systemd.cil

**Comment by poettering** [comment]
@mgrepl i either need a patch from you or somebody else for the tmpfs mount option solution, or I can just merge #259. Just let me now which option you prefer!

**Comment by ghost** [comment]
@mgrepl please just ACK this. It is a relatively small change. We can see in rawhide how it goes (it should not affect anything unless we adjust the security policy to take advantage of this)

**Comment by mgrepl** [comment]
@doverride I am able to get it working (pam_selinux+systemd-user) but of course with issues which I wrote above. It looks you have a different policy. 

$ ls -Z /usr/lib/systemd/systemd

**Comment by ghost** [comment]
@mgrepl , yes i do have a different policy. For one my systemd_t domain is strictly confined as opposed to your init_t domain which is pretty much unconfined.

**Comment by mgrepl** [comment]
But you still need to define entrypoint permission for pam_selinux.

**Comment by ghost** [comment]
@mgrepl Good point. I do not have any unconfined users so you may or may not have identified a bug in my policy :)

**Comment by ghost** [comment]
@mgrepl oh no, It is fine in my policy. all executable file types are entry points to the unconfineduser_t domain automatically.

`(allow subject_unconfined_subject_type subject_entry_object_type (file (entrypoint)))`

**Comment by mgrepl** [comment]
We removed this functionality in Fedora.

So the problem remains, we would need to allow entrypoint on init_exec_t for userdomains.

**Comment by ghost** [comment]
@mgrepl it was implemented differently in Fedora i believe.

In Fedora _all_ file types were entry points to the unconfined_t domain i think. If i am correct this was removed after the setroubleshoot spat.

In my policy this only applies to subject_entry_objects

**Comment by ghost** [comment]
@mgrepl I might be wrong. Anyhow, yes our policies differ. It would not make sense for me to maintain my own policy if our policies were identical or close to identical.

**Comment by mgrepl** [comment]
@poettering Let me test #259. But it looks fine for me.

**Comment by mgrepl** [comment]
@doverride  How do you implement /run/user/<UID> labeling? unconfined_u vs. staff_u for example. 

Also do you want to have a different label for root vs. other users?

I see on my VM:

drwx------. 10 mgrepl mgrepl system_u:object_r:user_tmp_t:s0 200 Jun 19 12:17 1000

drwx------. 2 mgrepl mgrepl unconfined_u:object_r:config_home_t:s0  60 Jun 19 13:42 dconf
drwx--x--x. 2 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0     60 Jun 18 13:27 gdm
drwx------. 3 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0     60 Jun 18 13:27 gnome-shell
dr-x------. 2 mgrepl mgrepl system_u:object_r:fusefs_t:s0            0 Jun 18 13:27 gvfs
drwx------. 2 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0    120 Jun 18 13:27 keyring
drwx------. 2 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0     40 Jun 19 12:17 midori
drwx------. 2 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0     80 Jun 18 13:27 pulse
drwxr-xr-x. 2 mgrepl mgrepl unconfined_u:object_r:user_tmp_t:s0     80 Jun 18 13:27 systemd

Not sure if we can get it working easily as you think. 

But I have no problem to have the systemd patch.

**Comment by ghost** [comment]
I would probably extent `genhomedircon()` and use similar mechanism as the one used with the `HOME_DIR` keyword.

Example:

`RUNTIME_DIR/systemd(/.*)? gen_context(system_u:object_r:user_systemd_runtime_t,s0)`

Or something along those lines.

I my case my laptop shell script deals with labeling generation.

But, yes, this change does not force you to take advantage of it. If you want you can just keep the same solution for `/run/user` as you currently have. This should not affect that in any way.

**Comment by ghost** [comment]
I haven' t really thought about `/run/user/0` mainly because I generally do not login as root. I probably would treat it like any other `UID` in this case for compatibility.

**Comment by ghost** [comment]
@mgrepl The `XDG basedir spec` does not make it any easier, since it leaves room open for interpretation as to whether to use `UID` or names.

In some distributions (Gentoo) name is used (/run/user/joe) as opposed to `UID` (/run/user/1000).

When names are used then the solution for us seems easier. We can just use the existing `USER` keyword and `libsemanage` will take care of the remainder.

The problem is that `systemd-logind` does not use names, instead it uses `UID`.

The reasoning goes:

"""using a UID instead of a name avoids the problem of needing NSS"""
"""it used to be the username, but it was changed because of the above (there were some services that broke due to the NSS requirement)"""

Some others then ask the question:

"""the problem of needing NSS? but dont you need NSS when you login anyway?"""

In the end, I suppose, we probably need to be able to support both if anything.

**Comment by ghost** [comment]
@mgrepl We'll probably work to extent `libsemanage` to introduce handling of `USERID`, then you can just use something like:

`/var/run/user/USERID(/.*)? gen_context(system_u:object_r:user_tmp_t,s0)`

Although then you can probably no longer rely on the substitution in `file_contexts.subs_dist` (if there are any). I am also wondering whether the introduction of `USERID` in `genhomedircon` will be sufficient. Especially with regard to the potential requirements for poly instantiation `(pam_namespace)`

**Comment by grawity** [comment]
As far as I can remember, the change was meant to make it easier for sssd, rpc.gssd, cifs-upcall to find the user's Kerberos tickets. (They run in a service environment, so they need to hardcode the expected path.)

So, in a sense, `/run/user/$UID` is already in use as an API by other projects, and Gentoo really should follow.

**Comment by ghost** [comment]
I tested https://github.com/systemd/systemd/pull/259 with v221/Fedora Rawhide. It is not good enough.

It looks like `lsetfilecon()` behaves like `restorecon` (without `-F`)

We need instead something that behaves like `restorecon -F` so that it also resets the context if only the identity or security level differs.

I am not sure which SELinux interface is suitable for that. It may be `selinux_lsetfilecon_default()`

I was not able to try that out since `systemd` does not build on `Fedora Rawhide` due to dependency issues `(yum-builddep systemd.spec fails)`

**Comment by ghost** [comment]
@teg @poettering @mgrepl @dwalsh

`label_fix()` is not sufficient in my view. We would probably need something like `label_fix_force()` for this situation. It should look pretty much the same as `label_fix()` i suppose with the exception of `s/lsetfilecon/selinux_lsetfilecon_default`

I haven't tested that theory and practice because systemd 221 does not build on Fedora Rawhide by default. Should it not be build-able on Rawhide? Maybe the `buildrequires` could be adjusted from `==` to `>=` or something along those lines?

**Comment by teg** [comment]
@doverride I don't really know this stuff very well, care to explain why label_fix() is not sufficient?

**Comment by ghost** [comment]
`label_fix()` `(lsetfilecon())` only resets the security context specified by the security policy _if_ the type identifier of the target object differs from what is specified by the security policy.

A security context is made up of various identifiers delimited by `:`, and not just a type identifier. The `identity` and `security level` identifiers are meant to be "variable". This means that they should not be reset by default since generally the policy does not know what these identifiers should be.

This means that if you for example do not necessarily wish to reset the type identifier of `/run/user/$UID` but only the "variable identifiers"  like identity, and security level, you need to "force" a reset of the security context. The `lsetfilecon()` interface does not forcibly reset the context. The `selinux_lsetfilecon_default()` interface resets the context of the object regardless of type.

The nature of the `/run/user/$UID` object requires that we reset the context forcibly. (we need to use `selinux_lsetfilecon_default()` instead of `lsetfilecon()` to force a reset.)

`lsetfilecon()` is the `libselinux` equivalent of running `restorecon`, whereas `selinux_lsetfilecon_default()` is the equivalent of running `restorecon -F`

In the case of `/run/user/$UID` we need the latter.

Simply put: `lsetfilecon()` only resets the type identifier, whereas `selinux_lsetfilecon_default()` resets the whole security context.

Example SecurityContext:
`Identity:Role:Type:SecurityLevel`

I hope this made a little sense to you.

**Comment by mgrepl** [comment]
@doverride Are you sure you have correctly defined labeling for /run/user/<UID>? What does matchpathcon say? There is selabel_lookup_raw() which should return a correct label which is passed to lsetfilecon().

**Comment by ghost** [comment]
@mgrepl yes I am pretty sure. matchpathcon returns the proper context. It is just not being applied by systemd-logind

We need to make sure that all security identifiers are reset on $XDG_RUNTIME_DIR, and not just the type identifier.

**Comment by poettering** [comment]
OK, so what is this bug about now? I don't understand much of what is written here. One of the issues was fixed with #259 afaics. Given that this issue is now very confusing and discusses three different issues at once, I'll now close it. Please open separate issues for the remaining issues, and consisely explain what is missing, and even better, supply a patch.

I rely on contributes patches for SELinux support, as I do not understand SELinux sufficiently to do this on my own.

Anyway, closing, please open separate issues, thanks.

**Comment by ghost** [comment]
Yes forget about it. This is a very specific issue that pretty much only affects the "MLS" crowd. Basically MLS enforces a model based on "no read up and no write down" to enforce confidentiality. so there is a range of Security levels of lets say 1 3 4 5 6 7. and processes get associated with a current level and optionally a clearance. Logind can't handle this so it creates $XDG_RUNTIME_DIR with the lowest level. Then when a user logs in with a level higher then the lowest level he is not allowed to use $XDG_RUNTIME_DIR because he can't "write down", and add entries to the $XDG_RUNTIME_DIRECTORY associated with the lowest level.

So if you for example tell SELinux to associate level 2 with the login shell and systemd-user instance then nothing can be created in there and it becomes useless.

The suggested fix above does not even deal with the complete issue. The would also need to be some support for poly instantiation. E.g. if a user has access to a clearance then he might need to login on varios security levels and for each level a $XDG_RUNTIME_DIR needs to be instantiated.

I do not use MLS myself, plus i do not have what it takes to provide patches to solve this issue. Therefore, let it be. Maybe the MLS community will one day step up if they believe it is something worth implementing
